plugins/outputs/firehose/firehose.go

```diff
@@ -0,0 +1,268 @@
+
+  ## debug will show upstream aws messages.
+  debug = false
+`
+
+func (f *FirehoseOutput) SampleConfig() string {
+	return sampleConfig
+}
+
+func (f *FirehoseOutput) Description() string {
+	return "Configuration for the AWS Firehose output."
+}
+
+func (f *FirehoseOutput) Connect() error {
+	// We attempt first to create a session to Firehose using an IAMS role, if that fails it will fall through to using
+	// environment variables, and then Shared Credentials.
+	if f.Debug {
+		log.Printf("E! firehose: Building a session for connection to Firehose in %+v", f.Region)
+	}
+
+	credentialConfig := &internalaws.CredentialConfig{
+		Region:    f.Region,
+		AccessKey: f.AccessKey,
+		SecretKey: f.SecretKey,
+		RoleARN:   f.RoleARN,
+		Profile:   f.Profile,
+		Filename:  f.Filename,
+		Token:     f.Token,
+	}
+	configProvider := credentialConfig.Credentials()
+
+	// we simply create the skeleton here. AWS doesn't attempt
+	// any connection yet, so we don't know if an error will happen.
+	svc := firehose.New(configProvider)
+	f.svc = svc
+	return nil
+}
+
+func (f *FirehoseOutput) Close() error {
+	return nil
+}
+
+func (f *FirehoseOutput) SetSerializer(serializer serializers.Serializer) {
+	f.serializer = serializer
+}
+
+// writeToFirehose uses the AWS GO SDK to write batched records to firehose and
+// queue the failed writes to the errorBuffer of f
+func writeToFirehose(f *FirehoseOutput, b [][]byte, submitAttemptCount int64) {
+	start := time.Now()
+	dataSize := 0 // track the size of the record in bytes
+	var values []byte
+
+	// take metrics and flatten into a record
+
+	if f.EnableGzipCompression == true {
+		var buf bytes.Buffer
+		zw := gzip.NewWriter(&buf)
+
+		for _, line := range b {
+			_, err := zw.Write(line)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+
+		if err := zw.Close(); err != nil {
+			log.Fatal(err)
+		}
+
+		values = buf.Bytes()
+		dataSize = len(values)
+	} else {
+		for _, line := range b {
+			values = append(values, line...)
+			dataSize = dataSize + len(line)
+		}
+	}
+
+	// todo.. check that we are not exceeding the max record size of 1MB to firehose
+	// though.. we'd be really impressed if any single telegraf metric write is >1MB. That's a lot.
+	// for now, i've updated the docstring to ask that users adjust metric_batch_size and flush_interval
+	// if they run into this.  A proper solution here would need to predict and (if needed) move lines
+	// around to get as close to 1MB as possible.. the level of effort is high enough to postpone for now.
+
+	d := firehose.Record{
+		Data: values,
+	}
+
+	r := []*firehose.Record{}
+	r = append(r, &d)
+
+	batchInput := &firehose.PutRecordBatchInput{}
+	batchInput.SetDeliveryStreamName(f.DeliveryStreamName)
+	batchInput.SetRecords(r)
+
+	// attempt to send data to firehose
+	resp, err := f.svc.PutRecordBatch(batchInput)
+
+	// if we had a failure, log it and enqueue the request for next time
+	if (aws.Int64Value(resp.FailedPutCount) > 0) || (err != nil) {
```

Comment:
If you return an error from this function, Telegraf will automatically retry the same batch next flush.  So if this is believed to be a transient error then just return the err.  If it is a permanent error, then you have no choice but to log and return nil to discard the data to avoid permanently being stuck.